shared/shot.js

```diff
@@ -0,0 +1,726 @@
+    this._backend = backend;
+    this._id = id;
+    this.origin = attrs.origin || null;
+    this.fullUrl = attrs.fullUrl || null;
+    if ((! attrs.fullUrl) && attrs.url) {
+      console.warn("Received deprecated attribute .url");
+      this.fullUrl = attrs.url;
+    }
+    this.docTitle = attrs.docTitle || null;
+    this.userTitle = attrs.userTitle || null;
+    this.createdDate = attrs.createdDate || Date.now();
+    this.favicon = attrs.favicon || null;
+    this.siteName = attrs.siteName || null;
+    this.images = [];
+    if (attrs.images) {
+      this.images = attrs.images.map(
+        (json) => new this.Image(json));
+    }
+    this.openGraph = attrs.openGraph || null;
+    this.twitterCard = attrs.twitterCard || null;
+    this.documentSize = attrs.documentSize || null;
+    this.fullScreenThumbnail = attrs.fullScreenThumbnail || null;
+    this.abTests = attrs.abTests || null;
+    this._clips = {};
+    if (attrs.clips) {
+      for (let clipId in attrs.clips) {
+        let clip = attrs.clips[clipId];
+        this._clips[clipId] = new this.Clip(this, clipId, clip);
+      }
+    }
+
+    for (let attr in attrs) {
+      if (attr !== "clips" && attr !== "id" && this.REGULAR_ATTRS.indexOf(attr) === -1 && this.DEPRECATED_ATTRS.indexOf(attr) === -1) {
+        throw new Error("Unexpected attribute: " + attr);
+      } else if (attr === "id") {
+        console.warn("passing id in attrs in AbstractShot constructor");
+        console.trace();
+        assert(attrs.id === this.id);
+      }
+    }
+  }
+
+  /** Update any and all attributes in the json object, with deep updating
+      of `json.clips` */
+  update(json) {
+    let ALL_ATTRS = ["clips"].concat(this.REGULAR_ATTRS);
+    assert(checkObject(json, [], ALL_ATTRS), "Bad attr to new Shot():", Object.keys(json));
+    for (let attr in json) {
+      if (attr == "clips") {
+        continue;
+      }
+      if (typeof json[attr] == "object" && typeof this[attr] == "object" && this[attr] !== null) {
+        let val = this[attr];
+        if (val.asJson) {
+          val = val.asJson();
+        }
+        if (! deepEqual(json[attr], val)) {
+          this[attr] = json[attr];
+        }
+      } else if (json[attr] !== this[attr] &&
+          (json[attr] || this[attr])) {
+        this[attr] = json[attr];
+      }
+    }
+    if (json.clips) {
+      for (let clipId in json.clips) {
+        if (! json.clips[clipId]) {
+          this.delClip(clipId);
+        } else if (! this.getClip(clipId)) {
+          this.setClip(clipId, json.clips[clipId]);
+        } else if (! deepEqual(this.getClip(clipId).asJson(), json.clips[clipId])) {
+          this.setClip(clipId, json.clips[clipId]);
+        }
+      }
+    }
+
+  }
+
+  /** Returns a JSON version of this shot */
+  asJson() {
+    let result = {};
+    for (let attr of this.REGULAR_ATTRS) {
+      var val = this[attr];
+      if (val && val.asJson) {
+        val = val.asJson();
```

Comment:
The `toJSON` method is the standard way of doing this, and is automatically called by `JSON.stringify`